### PR TITLE
Lambda -1 implementation:

### PR DESCRIPTION
**Pull request by SinaSina92**
* The Lambda -1 interface can be set under the [simulation] section in the TOML file (e.g., lambda_minus_one = -34).

* The starting condition can be either L or R in the 000 ensemble.

* All 0- paths hitting the Lambda -1 interface will be accepted if they are valid (0- paths can be RML, RMR, LMR, LML).

* For RETIS zero swap, the swap move will be rejected if the 0- path is RML or LML (with status 0-L).

### CODE REVIEW

File: infretis/core/tis.py [hunk 1]
```diff
@@ -900,18 +905,25 @@ def retis_swap_zero(
...
+    if (
+        (
+            "L" not in set(ens_set0["start_cond"])  # if not lambda_minus_one, 0- start_cond is ['R']
+            and "L" in path0.check_interfaces(ens_set0["interfaces"])[:2]
+        ) 
+        or (
+            "L" in set(ens_set0["start_cond"])      # if lambda_minus_one, 0- start_cond is ['L', 'R']
+            and "R" in set(ens_set0["start_cond"]) 
+            and "L" in path_old0.check_interfaces(ens_set0["interfaces"])[1:2] # reject 0- RML and LML swap with 0+
+        )
```
Comment (dz24) [inline infretis/core/tis.py:918]: I think we can simply do an early rejection before doing any MD steps at all in the case that path_old0 is either RML or LML

```python3
    # if lambda_minus_one, reject early if path_old0
    if set(ens_set0["start_cond"]) == set(["L", "R"]):
        if path_old0.check_interfaces(ens_set0["interfaces"])[1] == "L":
            return False, [path_old0, path_old1], "0-L"
```
e.g. like this and put it on before the line `    # 1. Generate path for [0^-] from [0^+]:`

**Comment by dz24** [comment]
You can also put the following pytests at the end of the `test_zero_swaps` function, in file `infretis/test/tis/test_tis.py,`

```python3
    # test the lambda_minus_one feature for retis_swap_zero
    if zero_swap_move.__name__ == "retis_swap_zero":
        pathdir = PosixPath(tmp_dir / "zero_swap_move")
        pathdir.mkdir()
        turtle.exe_dir = str(pathdir.resolve())
        intf_minus = (-1.05, -1.02, -0.99)
        picked[-1]["ens"]["interfaces"] = intf_minus
        picked[-1]["ens"]["start_cond"] = ["L", "R"]

        success, [new_path0, new_path1], status = zero_swap_move(
            picked, engines
        )

        # LMR path should be generated
        start, end, middle,  _ = new_path0.check_interfaces(intf_minus)
        assert start == "L"
        assert end == "R"
        assert middle == "M"
```

you can see that the test passes by running `pytest test_tis.py ` in the terminal and in that file directory

**Comment by dz24** [comment]
I guess for now we disable lambda -1 if quantis is run @lukasbaldauf, so if you may in `infretis/infretis/setup.py`, add 

```python3
    quantis = config["simulation"]["tis_set"].get("quantis", False)
    lambda_minus_one = config["simulation"].get("lambda_minus_one", False)

    if quantis and lambda_minus_one:
        raise TOMLConfigError("Cannot run quantis with lambda_minus_one!")
```
in the `check_config` function

**Comment by dz24** [comment]
to pass the github checks, in the base infretis folder try

`black infretis`
`ruff check infretis`

and look at the required fixes in the terminal

**Comment by lukasbaldauf** [comment]
> I guess for now we disable lambda -1 if quantis is run 

Yes I think that’s okay, I feel they wouldn’t be used in combination too much, since it should be used speed up the [0-] paths.

Also, in the check_config, maybe we could add a check that if lambda_minus_one is set, it should be less than interface0.

**Comment by SinaSina92** [comment]
Thanks guys for the comments. I added all the suggestions. Also, ran black and ruff and fixed the issues.

**Comment by lukasbaldauf** [comment]
Thanks for fixing these things!

I hope I'm not too picky now, but I have 2 other teeny-tiny things. I would change it myself but I don't think I can push to your fork.

-----
**1:** Try using `.get()` only once, as we had some issues with this earlier:
```python
lambda_minus_one = self.config["simulation"].get(
```
I think we could set the keyword only once in `setup_config` in infretis/setup.py and then use something like
```python
lambda_minus_one = self.config["simulation"]["lambda_minus_one"]
```
in the repex.py functions (without usining `.get`). That can avoid some errors.

---- 
**2:** Not really a problem: in the .toml, consider moving the `lambda_minus_one` keyword to the [simulation][simulation.tis_set] section.

Everything else looks neat!

**Comment by SinaSina92** [comment]
No problem!

I had to define a property in `repex.py` so that I could use `lambda_minus_one` everywhere. Now, we use `.get()` once in `setup_config` and once in `repex.py`. It has also been moved to `simulation.tis_set`.

Let me know if there’s anything else you’d like to change.

**Comment by lukasbaldauf** [comment]
It was already very nice before!

I think my suggestions were a bit unclear, so I modified it now with what I had in mind.

.Do you think this would work?

**Comment by SinaSina92** [comment]
Everything looks fine. There was just a small issue with inconsistency in the default value for `lambda_minus_one`. In some places, it was `None`, and in others, it was `False`. I fixed it, and now it is set to `False` everywhere.

**Comment by lukasbaldauf** [comment]
Yea sorry for that, it was me that mixed the False up in there. I caught one more default one.

One more thing, how do you go about analyzing the results? With the regular WHAM in inftools i get an error since the LL paths have a max op less than interface. Or is that maybe something yet to fix?

**Comment by SinaSina92** [comment]
Yeah, exactly, I got that too when I was using the WHAM. I don’t have a solution for it right now, but I was talking with Titus about it last month, and I’ll try to fix it by the end of January. For now, we need to delete the 0-LML path from infretis.data.

By the way, I’ve written an order parameter for my case that uses MDAnalysis. It takes two groups from the TOML file and calculates the ZCOM distance between them as lambda. Do you think it would be easy to add that to the main branch? I was wondering if using MDAnalysis might be an issue. You’re probably aware of the problems with MDAnalysis because the TPR parser version changes with every GROMACS release, which then requires updates to MDAnalysis.

**Comment by dz24** [comment]
> By the way, I’ve written an order parameter for my case that uses MDAnalysis. It takes two groups from the TOML file and calculates the ZCOM distance between them as lambda. Do you think it would be easy to add that to the main branch? I was wondering if using MDAnalysis might be an issue. You’re probably aware of the problems with MDAnalysis because the TPR parser version changes with every GROMACS release, which then requires updates to MDAnalysis.

I think maybe adding it to a new folder called "order parameters" in [infentory](https://github.com/infretis/infentory) as some sort of template would be better, so there would be less to maintain in this repo?

One single trivial test fails when running `pytest test` (all the written tests are ran). The variable `lambda_minus_one` is now always saved in `restart.toml`, while the corresponding test files in `infretis/test/simulations/data/10steps_wf`, `10steps_wf` and `30steps_wf` does not have them. So if you could add them, e.g. 

```
--- a/test/simulations/data/10steps_wf/restart.toml
+++ b/test/simulations/data/10steps_wf/restart.toml
@@ -68,6 +68,7 @@ allowmaxlength = false
 zero_momentum = false
 n_jumps = 6
 quantis = false
+lambda_minus_one = false
 accept_all = false
```

Then I'll merge this PR.

(Personally I am not sure if this is the best way to test (and always have to add the new keywords in the test templates, but oh well)).

**Review by dz24 — APPROVED** [review]
LGTM

**Pull request merged**